### PR TITLE
LPS-192181

### PR DESCRIPTION
**Pull request by gabrielwas**

### CODE REVIEW

**Comment by MarinhoFeliphe** [comment]
ci:test:relevant

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectDefinitionServiceTest.java [hunk 1]
```diff
@@ -168,6 +168,26 @@ public void testUpdateCustomObjectDefinition() throws Exception {
...
+			Assert.fail();
+		}
+		catch (PrincipalException.MustHavePermission principalException) {
+			String message = principalException.getMessage();
+
+			Assert.assertTrue(
+				message.contains(
+					"User " + _user.getUserId() +
+						" must have UPDATE permission for"));
+		}
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectDefinitionServiceTest.java:185]: @gabrielwas here we can use assertFailure as well, for example:

```suggestion
		AssertUtils.assertFailure(
			PrincipalException.MustHavePermission.class,
			"User " + _user.getUserId() + " must have UPDATE permission for",
			() -> _testUpdateRootObjectDefinitionId(_adminUser, _user));
```

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 2]
```diff
@@ -992,6 +997,30 @@ private ObjectRelationship _updateObjectRelationship(
...
+				"Object relationship must be one to many to be an edge of a " +
+					"root context");
+		}
+
+		if (objectDefinitionId1 == objectDefinitionId2) {
+			throw new ObjectRelationshipEdgeException(
+				"Object relationship must not be a self-relationship to be " +
+					"an edge of a root context");
+		}
+	}
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1022]: @gabrielwas here we can pass the objectRelationship as parameter, so we can do:

```suggestion
	private void _validateEdge(boolean edge, ObjectRelationship objectRelationship)
		throws ObjectRelationshipEdgeException {

		if (!edge) {
			return;
		}

		if (!Objects.equals(
				objectRelationship.getType(), ObjectRelationshipConstants.TYPE_ONE_TO_MANY)) {

			throw new ObjectRelationshipEdgeException(
				"Object relationship must be one to many to be an edge of a " +
					"root context");
		}

		if (objectRelationship.isSelf()) {
			throw new ObjectRelationshipEdgeException(
				"Object relationship must not be a self-relationship to be " +
					"an edge of a root context");
		}
	}
```

there is another validation that I believe we can do here, we need to check if the father is a unmodifiable system object, if so, we need to throw an exception.

Thanks!

**Review by MarinhoFeliphe — CHANGES_REQUESTED** [review]
@gabrielwas feel free to get in touch if you disagree with my suggestions,

nice changes by the way! :tada: 

Thank!

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java [hunk 3]
```diff
@@ -1891,6 +1909,26 @@ private void _validatePluralLabel(Map<Locale, String> pluralLabelMap)
...
+			long objectDefinitionId, long rootObjectDefinitionId)
+		throws PortalException {
+
+		ObjectDefinition objectDefinition =
+			objectDefinitionPersistence.fetchByPrimaryKey(
+				rootObjectDefinitionId);
+
+		if (objectDefinition == null) {
+			throw new NoSuchObjectDefinitionException();
+		}
```
[the remark this answers is not in the document]
Comment (gabrielwas) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java:1922]: Good catch!

**Comment by MarinhoFeliphe** [comment]
@gabrielwas do you think we should add the FF in other places like the validation methods for examploe?

about [this commit](https://github.com/liferay-objects/liferay-portal/pull/2418/commits/cfe0e1acb118b89f3543dc8397bd554753b77cca) I'm not sure if you use the proper LPS in the commit message

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java (hunk 2 above)
Comment (gabrielwas) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1022]: I did it at first, but then I notice the pattern of the validation methods is passing the parameters (and not the major entity).

**Comment by gabrielwas** [comment]
ci:stop

**Comment by gabrielwas** [comment]
ci:test:sf

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarinhoFeliphe) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java]: https://github.com/liferay-objects/liferay-portal/pull/2418/commits/8b8a39ac3a331df06bda3d4ebeb4f76a983106dc

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java [hunk 4]
```diff
@@ -1891,6 +1909,26 @@ private void _validatePluralLabel(Map<Locale, String> pluralLabelMap)
 		}
 	}
 
+	private void _validateRootObjectDefinitionId(
+			long objectDefinitionId, long rootObjectDefinitionId)
+		throws PortalException {
```
[the remark this answers is not in the document]
Comment (MarinhoFeliphe) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java]: https://github.com/liferay-objects/liferay-portal/pull/2418/commits/8b8a39ac3a331df06bda3d4ebeb4f76a983106dc

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java [hunk 5]
```diff
@@ -408,6 +421,37 @@ private static ObjectDefinition _addSystemObjectDefinition(
...
+		_objectRelationshipLocalService.addObjectRelationship(
+			TestPropsValues.getUserId(),
+			_objectDefinition1.getObjectDefinitionId(),
+			_objectDefinition2.getObjectDefinitionId(), 0,
+			ObjectRelationshipConstants.DELETION_TYPE_PREVENT,
+			LocalizedMapUtil.getLocalizedMap(RandomTestUtil.randomString()),
+			objectRelationshipName,
+			ObjectRelationshipConstants.TYPE_ONE_TO_MANY);
+
+		ObjectField objectField3 = _objectFieldLocalService.getObjectField(
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:438]: @gabrielwas we don't have conflicts here, so we can rename to objectField

**Review by MarinhoFeliphe — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectDefinitionLocalServiceTest.java [hunk 6]
```diff
@@ -1555,34 +1555,30 @@ public void testUpdateExternalReferenceCode() throws Exception {
...
+			() -> _objectDefinitionLocalService.updateRootObjectDefinitionId(
+				objectDefinition1.getObjectDefinitionId(),
+				objectDefinition2.getObjectDefinitionId()));
 
-		_objectDefinitionLocalService.deleteObjectDefinition(
-			customObjectDefinition1);
-		_objectDefinitionLocalService.deleteObjectDefinition(
-			customObjectDefinition2);
+		_objectDefinitionLocalService.deleteObjectDefinition(objectDefinition1);
+		_objectDefinitionLocalService.deleteObjectDefinition(objectDefinition2);
```
[the remark this answers is not in the document]
Comment (gabrielwas) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectDefinitionLocalServiceTest.java:1581]: @MarinhoFeliphe We're not using @DataGuard in this test class. So I believe they're not deleted after each execution. And the other methods are deleting the object definitions.

**Comment by gabrielwas** [comment]
ci:test:relevant

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectDefinitionLocalServiceTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (MarinhoFeliphe) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectDefinitionLocalServiceTest.java:1581]: got it, thanks!

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java (hunk 5 above)
Comment (MarinhoFeliphe) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java]: https://github.com/liferay-objects/liferay-portal/pull/2418/commits/df40f2c3dcd095447c077e0fdab4385957f49c66

**Review by MarinhoFeliphe — APPROVED** [review]
LGTM!! :tada:

**Comment by MarinhoFeliphe** [comment]
ci:test:sf

**Comment by MarinhoFeliphe** [comment]
ci:test:object

**Comment by gabrielwas** [comment]
ci:test:relevant

**Comment by gabrielwas** [comment]
ci:test:relevant

**Comment by gabrielwas** [comment]
resent here https://github.com/liferay-objects/liferay-portal/pull/2434

**Pull request closed without merging**